Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/concavity.h

```diff
@@ -0,0 +1,162 @@
+
+            TriangleMesh cluster;
+            CGAL::copy_face_graph(filtered_mesh, cluster);
+
+            Concavity concavity(cluster, m_traits);
+            return concavity.compute();
+        }
+
+        /**
+         * Computes concavity value of the whole mesh.
+         */
+        double compute()
+        {
+            CGAL_assertion(!CGAL::is_empty(m_mesh));
+
+            TriangleMesh conv_hull;
+            std::vector<Point_3> pts;
+
+            if (CGAL::num_vertices(m_mesh) <= 3) return 0;
+
+            // extract the list points of the mesh
+            BOOST_FOREACH(vertex_descriptor vert, vertices(m_mesh))
+            {
+                pts.push_back(get(CGAL::vertex_point, m_mesh)[vert]);
+            }
+
+            // compute convex hull
+            CGAL::convex_hull_3(pts.begin(), pts.end(), conv_hull); 
+            
+            return compute(vertices(m_mesh), conv_hull);
+        }
+
+        /**
+         * Constructs list of vertices from the list of faces and computes concavity value with the convex hull provided.
+         * Faces list is a subset of all faces in the mesh.
+         */
+        double compute(const std::vector<face_descriptor>& faces, const TriangleMesh& conv_hull)
+        {
+            std::unordered_set<vertex_descriptor> pts;
+
+            BOOST_FOREACH(face_descriptor face, faces)
+            {
+                BOOST_FOREACH(vertex_descriptor vert, vertices_around_face(halfedge(face, m_mesh), m_mesh))
+                {
+                    pts.insert(vert);
+                }
+            }
+
+            return compute(std::make_pair(pts.begin(), pts.end()), conv_hull);
+        }
+
+        /**
+         * Computes concavity value projecting vertices from a list onto a convex hull.
+         * Vertices list a subset of all vertices in the mesh.
+         */
+        template <class iterator>
+        double compute(const std::pair<iterator, iterator>& verts, const TriangleMesh& conv_hull)
+        {
+            // compute normals if normals are not computed
+            compute_normals();
+
+            // construct AABB for fast computations of intersections between ray and convex hull
+            AABB_tree tree(faces(conv_hull).begin(), faces(conv_hull).end(), conv_hull);
+
+            double result = 0;
+
+            // compute intersections and select the largest segment length
+            BOOST_FOREACH(vertex_descriptor vert, verts)
+            {
+                Point_3 origin = get(CGAL::vertex_point, m_mesh)[vert];
+                Ray_3 ray(origin, m_normals_map[vert]);
+                
+                Ray_intersection intersection = tree.first_intersection(ray);
+                if (intersection)
+                {
+                    const Point_3* p =  boost::get<Point_3>(&(intersection->first));
+                    if (p)
+                    {
+                        result = std::max(result, CGAL::squared_distance(origin, *p));
+                    }
```

Comment:
No intersection. Actually, it should never happen because each point is projected onto the convex hull which contains them inside, so I think there is no need to add infinite concavity value.

Reply:
Hmm, but it's checked in the previous if statement. This one should be removed.

Reply:
The intersection of a ray and a triangle is either a segment or a point. If the normal is noisy, you might endup having a segment as intersection point. In such a case I suggest to simply ignore the vertex contribution.

Reply:
Ok, then I think I did fine at the beginning (checked for intersection and then checked if it's a point).

Reply:
then you need to update the code